fix: NetworkAnimator does not work using the server authoritative model [MTT-3745]

NoelStephensUnity opened the pull request:
The more recent NetworkAnimator adjustments did not support server authoritative mode.
This PR makes NetworkAnimator default to server authoritative mode and provides the ability for users to derive from NetworkAnimator, override OnIsServerAuthoritative, and return false to set the NetworkAnimator into owner/client authoritative mode.

[MTT-3745](https://jira.unity3d.com/browse/MTT-3745)

## Changelog
- Fixed: `NetworkAnimator` could not run in the server authoritative mode.

## Testing and Documentation
- Includes integration test updates

NoelStephensUnity committed:
fix

Added server authoritative mode for NetworkAnimator.
Added the ability to toggle between server and client authoritative mode (from the server only).

NoelStephensUnity committed:
test manual

Updating the manual test to have both server and owner authoritative modes.

NoelStephensUnity committed:
test

MTT-3745
This includes adjustments to the NetworkAnimator tests that now also validates running in server authoritative mode.

NoelStephensUnity committed:
Merge branch 'develop' into fix/networkanimator-server-authoritative

NoelStephensUnity committed:
Update CHANGELOG.md

MTT-3745

NoelStephensUnity committed:
style

whitespace fixes.

NoelStephensUnity committed:
style

adding a tooltip and updated comments

NoelStephensUnity committed:
test

removing commented out code.

SamuelBellomo commented on com.unity.netcode.gameobjects/Components/NetworkAnimator.cs:
```diff
@@ -209,6 +208,42 @@ public Animator Animator
 
+        [Tooltip("When enabled, the NetworkAnimator will operate in server authoritative mode.")]
+        [SerializeField]
+        private bool m_IsServerAuthoritative;
```
Should this be set to true by default as to not create a breaking change?

NoelStephensUnity replied:
Good question.  If we want to make server authoritative the default that is easy enough to do.
Do we want users to default to the server authoritative model or client authoritative model?
_(both approaches have pro's and con's)_

0xFA11 replied:
+1 for making server-auth the default to not change the current default.

NoelStephensUnity replied:
NetworkAnimator follows the same pattern as NetworkTransform to make it client/owner authoritative.

SamuelBellomo commented:
Tested real quick in boss room and this does solve our problem with pre.10

0xFA11 commented on com.unity.netcode.gameobjects/Components/NetworkAnimator.cs:
```diff
@@ -209,6 +208,42 @@ public Animator Animator
+        /// Can be used to switch between server and owner authoritative modes during runtime.
+        /// Note: Only when the associated NetworkObject is spawned.
+        /// </summary>
+        public bool IsServerAuthoritative
```
to keep this API similar to `NetworkTransform`'s, I think we should go for something similar to (protected) `NetworkTransform::OnIsServerAuthoritative()` & (internal) `NetworkTransform::IsServerAuthoritative()`

0xFA11 commented on com.unity.netcode.gameobjects/Components/NetworkAnimator.cs:
```diff
@@ -872,7 +930,7 @@ public void SetTrigger(string triggerName)
         /// <param name="reset">If true, resets the trigger</param>
```
needs update, `reset` renamed as `setTrigger` and logic changed too.

0xFA11 committed:
Merge branch 'develop' into fix/networkanimator-server-authoritative

NoelStephensUnity committed:
update

Changing how owner authority is set in a way that is similar to NetworkTransform where a user will now derive from NetworkAnimator, override OnIsServerAuthoritative, and return false as opposed to true.
Made adjustments to the manual and integration tests to reflect this change.

NoelStephensUnity committed:
style

Updating the xml documentation parameter name for SetTrigger

NoelStephensUnity committed:
Merge branch 'develop' into fix/networkanimator-server-authoritative

0xFA11 committed:
Merge branch 'develop' into fix/networkanimator-server-authoritative
